src/python/zquantum/core/wavefunction.py

```diff
@@ -0,0 +1,149 @@
+
+    try:
+        return np.array(sympy_matrix, dtype=new_type).flatten()
+    except TypeError:
+        return np.array(sympy_matrix, dtype=object).flatten()
+
+
+class Wavefunction:
+    """
+    A simple wavefunction data structure that can
+    be used to calculate amplitudes of quantum states.
+
+    Args:
+        amplitude_vector: the initial amplitudes of the system,
+            can either be a NumPy ndarray or a SymPy Matrix
+    """
+
+    def __init__(
+        self, amplitude_vector: Union[np.ndarray, Matrix, List[complex]]
+    ) -> None:
+        if bin(len(amplitude_vector)).count("1") != 1:
+            raise ValueError(
+                "Provided wavefunction does not have a size of a power of 2."
+            )
+
+        try:
+            self._wavefunction = np.asarray(amplitude_vector, dtype=complex)
+        except TypeError:
+            self._wavefunction = Matrix(amplitude_vector)
+
+        self._check_sanity(self._wavefunction)
+
+    @property
+    def amplitudes(self):
+        return _cast_sympy_matrix_to_numpy(self._wavefunction, complex=True)
+
+    @property
+    def n_qubits(self):
+        return int(log2(len(self)))
+
+    @property
+    def free_symbols(self):
+        return _free_symbols(self._wavefunction)
+
+    @staticmethod
+    def _check_sanity(arr: Matrix):
+        """
+        Cases to watch out for:
+        #1 no free symbols --> must ensure unit probability
+        #2 free symbols --> must ensure numbers do not exceed unit probability already
+        """
+
+        def _calculate_probability_of_ground_entries(arr: Matrix) -> np.float64:
+            numbers = np.array(
+                [elem for elem in arr if _is_number(elem)], dtype=np.complex128
+            )
+            return np.sum(np.abs(numbers) ** 2)
+
+        probs_of_ground_entries = _calculate_probability_of_ground_entries(arr)
+        if not _free_symbols(arr):
+            if not np.isclose(probs_of_ground_entries, 1.0):
+                raise ValueError("Vector does not result in a unit probability.")
+        else:
+            if probs_of_ground_entries > 1.0:
+                raise ValueError(
+                    "Ground entries in vector already exceeding probability of 1.0!"
+                )
+
+    def __len__(self) -> int:
+        return len(self._wavefunction)
+
+    def __iter__(self):
+        return iter(self._wavefunction)
+
+    def __getitem__(self, idx):
+        return self._wavefunction[idx]
+
+    def __setitem__(self, idx, val):
+        old_val = self._wavefunction[idx]
+        self._wavefunction[idx] = val
+
+        try:
+            self._check_sanity(self._wavefunction)
+        except ValueError:
+            self._wavefunction[idx] = old_val
+
+            raise ValueError("This assignment violates probability unity.")
+
+    def __str__(self) -> str:
+        return self._wavefunction.__str__()
+
+    def __eq__(self, other) -> bool:
+        return self._wavefunction == other._wavefunction
```

Comment:
This implementation will fail if the wavefunction is compared to something that is not a wavefunction. It would be more pythonic to return  `False` instead of raising an exception in such cases.

Also, the `__eq__` method seems to not be covered by tests.  I imagine we should have positive and negative cases for wfs with/without free symbols and some negative cases for wfs with/without free symbols and a negative case for comparing wavefunction to a list.